### ISSUE TITLE
How to successfully avoid integer overflow?

### ISSUE THREAD

**Issue by matt-curtis** [libexpat/libexpat#371]
According to the [libexpat documentation](https://github.com/libexpat/libexpat/blob/master/expat/doc/reference.html#L1017):

> Note: the `len` argument passed to these functions
should be considerably less than the maximum value for an integer,
as it could create an integer overflow situation if the added
lengths of a buffer and the unprocessed portion of the previous buffer
exceed the maximum integer value. Input data at the end of a buffer
will remain unprocessed if it is part of an XML token for which the
end is not part of that buffer.

My understanding of this is that if there exists an XML fragment insufficient to generate a parse event, libexpat will keep hold of it, and on the next parse call, combine it with the new input. Together, it is possible that they might overflow the maximum value for a 32-bit integer.

If that understanding is correct, what is the recommended way to avoid this? In other words — what is the "considerably less"? libexpat doesn't expose how much data it has left unprocessed, so it can be difficult to feel safe from this error. I imagine hitting this problem would be rare (like a start or end tag with a very long name or a LOT of attributes), but I'd still like to know that if it _were_ to happen I'd be protected against it.

**Comment by hartwork** [comment]
Regarding the argument that you quoted above, the maximum should not exceed `INT_MAX / 2`.
With regard to https://trust-in-soft.com/objects-larger-than-ptrdiff_max-bytes/ , an inclusive limit of 65535 may be safer.

If you want to memory map the whole file — like xmlwf does by default — there will be a limit to file size in processing that can be exceeded by files in your file system, easily. For those bigger files, xmlwf offers argument `-r` — `disable memory-mapping and use normal file [r]ead IO calls instead`. You'll have to do something similar to xmlwf if you want to use *both* memory mapping and support arbitrary file size, or at least make some decision on that matter.

When you use a buffer of X bytes outside of Expat, copy bytes to it, and call `XML_Parse` with those bytes and `len=X`, Expat will invoke `XML_GetBuffer(..., len=X)` internally, and keep parsing the internal buffer (including any leftovers from previous rounds) while it can. That can be seen here:

https://github.com/libexpat/libexpat/blob/ed3484270fe45140e2d8858426472404b1f0ae6a/expat/lib/xmlparse.c#L1862-L1868

So the internal buffer will grow as large as it can and needs to be, the outer buffer (that you control) determines how many calls to `XML_Parse` you do yourself and how `malloc` will be called from inside `XML_Parse`. If you wanted control over the number of bytes allocated in advance during those internal call to `malloc`, you could use `XML_ParserCreate_MM` with a custom `XML_Memory_Handling_Suite` wrapper around `malloc`, `realloc` and `free` adding your own length block in front (similar to https://github.com/uriparser/uriparser/blob/51bdd9a3cb70c9020715cfe883f606cd5b995b7a/src/UriMemory.c#L167-L260 ).

In case you will benchmark the effect of (external) buffer size on parsing speed, please make sure to keep the disk/page buffer in mind and please share your results.

What do you think?

**Comment by matt-curtis** [comment]
Thank you for the comprehensive answer!

> Regarding the argument that you quoted above, the maximum should not exceed `INT_MAX / 2`.

To clarify, I'm working specifically with `XML_GetBuffer` and `XML_ParseBuffer`. So as the documentation dictates, for every new chunk of data I first call `XML_GetBuffer(chunkLength)`, then `XML_Parse(..., chunkLength, ...)`.

So I imagine the reasoning behind your recommendation here is if the maximum chunk size is set at `INT_MAX / 2`, then, when combined, the unprocessed data and incoming chunk together won't exceed `INT_MAX`. Correct?

> With regard to https://trust-in-soft.com/objects-larger-than-ptrdiff_max-bytes/ , an inclusive limit of 65535 may be safer.

Hmm, I'll have to dig into this one further when I have the opportunity.

I'm not memory mapping the file or using `XML_Parse` directly, but I do think the answer you've given provides guidance that is much lacking from the current documentation.

**Comment by hartwork** [comment]
> So I imagine the reasoning behind your recommendation here is if the maximum chunk size is set at `INT_MAX / 2`, then, when combined, the unprocessed data and incoming chunk together won't exceed `INT_MAX`. Correct?

If I think about it more, we need to rely on waterproof bounds checking inside the code for _any_ size of buffer because it can be more than two rounds accumulating: Imagine feeding a 10G document `<aa[..]aa/>` with a single attribute-less self-closing node in 4K chunks on a 32bit machine. That should exceed `INT_MAX` and if Expat doesn't detect it I would call it a bug.

> I'm not memory mapping the file or using `XML_Parse` directly, but I do think the answer you've given provides guidance that is missing from the current documentation.

Is there a particular place, where you would have expected this information? And which part in particular?

**Comment by matt-curtis** [comment]
> If I think about it more, we need to rely on waterproof bounds checking inside the code for _any_ size of buffer because it can be more than two rounds accumulating before we have enough text to be sure what it is and pass it to the right handlers.

You're right! I was assuming based on the documentation that there would only ever be two rounds of accumulated data. If there can be more than one, then I don't think there's any real way for an libexpat user to defend against an overflow scenario without the libexpat changing in someway.

> Is there a particular place, where you would have expected this information? And which part in particular?

Honestly I think all of it useful — it provides information on possible approaches depending on a user's needs. Since I'm using `XML_ParseBuffer`, what was specifically useful and insightful for _me_ was the information on how libexpat internally manages the buffer and the ideal max buffer input size (though considering that there can be more than 1 round of accumulated data in the backlog, that's not valid anymore...)

Right now, I'd have expected to find it in the same place I found the other disclaimers about using the `XML_Parse`, `XML_GetBuffer`, and `XML_Parse` functions:

https://github.com/libexpat/libexpat/blob/master/expat/doc/reference.html#L1004-L1023

**Comment by matt-curtis** [comment]
After further consideration, could `XML_GetCurrentByteIndex` be used to solve this problem? A libexpat user could compare the byte offset against how much data they know they've provided, and possibly avoid an overflow that way?

**Comment by hartwork** [comment]
I think as a user of Expat I would only worry about overflows in your own code and rely on Expat to catch all overflows in its own code. If there is a single place how does not do waterproof overflow prevention inside Expat it's a bug with security implications that needs to be fixed. (If you want to do some joint break-testing on Expat in that regard, please contact me offsite.)

I start to think that the buffer size is also maybe not that important regarding overall parsing performance. I would *not* pick something as small as 512 bytes but (a) 2^k number from 4096 to 32768 (if you care about https://trust-in-soft.com/objects-larger-than-ptrdiff_max-bytes/) or (c) the size of the file while below some limit (say 1-10 MiB).

What do you think?

**Comment by matt-curtis** [comment]
> I think as a user of Expat I would only worry about overflows in your own code and rely on Expat to catch all overflows in its code.

I agree; my suggestion above was intended as a possible workaround until libexpat is adjusted to account for this possibility.

> I start to think that the buffer size is also maybe not that important regarding overall parsing performance. I would _not_ pick something as small as 512 bytes but (a) 2^k number from 4096 to 32768 (if you care about https://trust-in-soft.com/objects-larger-than-ptrdiff_max-bytes/) or (c) the size of the file while below some limit (say 1-10 MiB).

Not sure I'm following you here. How does parsing performance fit into the worry about possible integer overflow?

**Comment by hartwork** [comment]
> I agree; my suggestion above was intended as a possible workaround until libexpat is adjusted to account for this possibility.

I only now get your `XML_GetCurrentByteIndex` idea and I think it might work. I'm not sure if it's worth the effort and you might reject some files that would parse fine otherwise, but depending on your use case and threat model maybe that's still a gain.

> > I start to think that the buffer size is also maybe not that important regarding overall parsing performance. I would _not_ pick something as small as 512 bytes but (a) 2^k number from 4096 to 32768 (if you care about https://trust-in-soft.com/objects-larger-than-ptrdiff_max-bytes/) or (c) the size of the file while below some limit (say 1-10 MiB).
> 
> Not sure I'm following you here. How does parsing performance fit into the worry about possible integer overflow?

If we assumed you left overflow catching to Expat, you'd still need to decide on some buffer size. I felt I hadn't really addressed that question earlier and this paragraph was meant to go into that direction.

**Comment by matt-curtis** [comment]
> I'm not sure if it's worth the effort and you might reject some files that would parse fine otherwise, but depending on your use case and threat model maybe that's still a gain.

Some additional context: I'm designing a wrapper API around libexpat, and not trying to address this felt like leaving a gaping hole in its design. I don't know how likely it is to actually run into this issue, but my general approach is to code and design defensively. At minimum, I'd like to be able to bail if an overflow is likely. Here's some psuedocode to demonstrate the idea:

```c
int inputChunkSize = ...;
int bufferSize = ...;
int bytesInBuffer = 0;
int previousByteIndex = 0;

while(parse()) {
	int currentByteIndex = get_current_byte_index();
	int expectedByteIndex = previousByteIndex + inputChunkSize;
	int numberOfBytesUnprocessed = expectedByteIndex - currentByteIndex;

	if(numberOfBytesUnprocessed > 0){
		int availableRemainingBufferSize = bufferSize - numberOfBytesUnprocessed;

		if(availableRemainingBufferSize > 0){
			// size next chunk accordingly
		} else {
			// bail
		}
	}
}
```

> If we assumed you left overflow catching to Expat, you'd still need to decide on some buffer size. I felt I hadn't really addressed that question earlier and this paragraph was meant to go into that direction.

I see. That affects only `XML_Parse`, since with `XML_ParseBuffer` the user is manually managing the buffer anyway, right? You mentioned earlier that the current implementation of `XML_Parse` grows the buffer as needed for the input, so I'm failing to see how it would be affected by any changes necessary for preventing overflow?

**Comment by hartwork** [comment]
> That affects only `XML_Parse`, since with `XML_ParseBuffer` the user is manually managing the buffer anyway, right?

I think it does affect  `XML_ParseBuffer` because you still need to pick a number for how many bytes to request for each call to `XML_GetBuffer`. No?

> You mentioned earlier that the current implementation of `XML_Parse` grows the buffer as needed for the input, so I'm failing to see how it would be affected by any changes necessary for preventing overflow?

Those thoughts were outside the overflow context and only about performance.

**Comment by matt-curtis** [comment]
> I think it does affect `XML_ParseBuffer` because you still need to pick a number for how many bytes to request for each call to `XML_GetBuffer`. No?

Hm... would my psuedocode example above solve that?

**Comment by hartwork** [comment]
> _If_ this understanding is correct, then this could be solved by modifying [`XML_GetBuffer`](https://github.com/libexpat/libexpat/blob/515f5d07c698d041cad5ec868002e6497be1c764/expat/lib/xmlparse.c#L1949) to detect whether or not resizing the buffer would cause an overflow or not, then emitting an error... or informing the user of what size input is safe.

To my understanding `XML_GetBuffer` does consider overflows regarding `int` but does not consider limits of `ptrdiff_t`. 

On x86_64 that _might_ be safe (but probably not in general):
```console
# uname -m
x86_64

# cat main.c 
#include <stdio.h>
#include <stddef.h>

int main() {
    printf("        sizeof(int): %ld\n", sizeof(int));
    printf("  sizeof(ptrdiff_t): %ld\n", sizeof(ptrdiff_t));
    return 0;
}

# gcc main.c 

# ./a.out 
        sizeof(int): 4
  sizeof(ptrdiff_t): 8
```

We cannot change the signature of `XML_GetBuffer` without breaking ABI and API compatibility.

> that `XML_GetBuffer` can be rewritten to avoid doing that arithmetic at all...

How?

> besides that, I'd like to know why the unprocessed and input buffers aren't maintained as separate buffers? AFAICT you could avoid a lot of arithmetic and memory copying that way.

I'm not sure I follow. If I have 500 bytes left to parse from the last round and 1000 bytes coming in now, passing 1500 bytes to the parser once as one single buffer seems reasonable to me. If it's two buffers thing get more complex. How did you imagine use of two buffers?

**Comment by matt-curtis** [comment]
> To my understanding `XML_GetBuffer` does consider overflows regarding `int` but does not consider limits of `ptrdiff_t`.

So, my opinion is: it's a hassle to try to implement overflow checking for every ptrdiff/int overflow. I think it better to avoid having to do pointer math and overflow checking at all.

This problem, _so far_ to me, seems to be introduced by the resizing math in trying to use one continuous buffer for parsing XML. Yes, it's more ergonomic to have a continuous buffer you can scan by doing addition or subtraction on a pointer, but I _think_ we can try a different approach that is faster than the resizing and copying `XML_GetBuffer` does, and create a similarly ergonomic abstraction for scanning over parts of a string (I have yet to look at those parts of the libexpat code, so that's an unsubstantiated claim atm, I'm going off of how I _think_ it'd look in theory. My plan is to double-check this.)

So assuming that the requirements are:

1. A way to move backwards and forwards over the bytes within the current segment of the XML string being parsed (currently my assumption is that this is being done within libexpat by some variation of `bufferStart + offset`) and
2. A way to keep track of an allocate buffers for both old, unprocessed XML, and new, incoming XML

I propose the following: all XML fragments/chunks be stored by libexpat as linked list of individual buffers. Here's a simple, incomplete implementation showing what this could look like (once again, my C is rusty, so I may be doing several things wrong):

```c
typedef struct XMLFragment {
	
	int size;
	void* buffer;
	void* nextFragment;
	
} XML_Fragment;

XML_Fragment *startingFragment;

void *XML_GetBuffer(int newFragmentSize) {
	//	Safety check on new fragment size:
	
	if(newFragmentSize <= 0) return 0;
	
	//	Try to allocate space for new fragment data
	
	void* newFragmentBuffer = malloc(newFragmentSize);
	
	if(!newFragmentBuffer) return 0;
	
	//	Create new XML_Fragment:
	
	XML_Fragment *newFragment = malloc(sizeof(XML_Fragment));
	
	if(!newFragment) return 0;
	
	newFragment->size = newFragmentSize;
	newFragment->buffer = newFragmentBuffer;
	newFragment->nextFragment = NULL;
	
	//	If this is the very first fragment we've received, assign it as the starting fragment:
	
	if(!startingFragment) {
		startingFragment = newFragment;
		
		return startingFragment->buffer;
	}
	
	//	Not the first fragment — grab the final fragment, then assign its nextBuffer to the a new fragment:
	
	XML_Fragment *currentFragment = startingFragment;
	
	while(true) {
		if(currentFragment->nextFragment) {
			currentFragment = currentFragment->nextFragment;
			
			continue;
		}
		
		//	Found final fragment! Append new fragment:
		
		currentFragment->nextFragment = newFragment;
		
		return newFragment->buffer;
	}
	
	return 0;
}

char XML_GetByte(int offset) {
	XML_Fragment *currentFragment = startingFragment;
	int fragmentByteCountSoFar = 0;
	
	while(currentFragment && fragmentByteCountSoFar <= offset) {
		int fragmentSize = currentFragment->size;
		
		if(offset < (fragmentByteCountSoFar + fragmentSize)){
			//	This fragment contains the given offset.
			
			int offsetWithinCurrentFragment = offset - fragmentByteCountSoFar;
			
			return ((char*)currentFragment->buffer)[offsetWithinCurrentFragment];
		}
		
		currentFragment = currentFragment->nextFragment;
	}
	
	return 0;
};
```

Pros:

1. No need to copy bytes around as new fragments arrive
2. No pointer arithmetic
3. Less overflow risk

Cons:

1. Slightly less ergonomic; `XML_GetByte` replaces the previous method of traversing the combination of all current fragments
2. Requires smattering of changes throughout overall codebase to replace old traversal method with new method
3. Time complexity of traversal changes: before, access was `O(1)`. With `XML_GetByte` it becomes, at max, roughly `O(n)`, where `n` is the current number of fragments in the list. I think this acceptable though, since it's unlikely that there will be many fragments around that there can be no event emitted for at a given time. If that's true, then this is on average closer to `O(1-2)`.

I think the benefits outweigh the negatives, but once again, this approach is somewhat naive — it does not take into consideration the entirety of the codebase's requirements; making this change may very well be non-trivial and not worth the effort.
